File: lib/dplua/dplog.lua
-- yxfangcs<yxfangcs@yeah.net> / 20241127 / 日志模块
--
local ffi = require("ffi")

ffi.cdef([[
typedef enum
{
    DPLOG_TA_SECOND = 0,
    DPLOG_TA_MILLIS = 1,
    DPLOG_TA_MICROS = 2,
} dplog_tsacc_e;

typedef enum
{
    DPLOG_L_DEBUG = 0,
    DPLOG_L_INFO = 1,
    DPLOG_L_NOTICE = 2,
    DPLOG_L_WARN = 3,
    DPLOG_L_ERROR = 4,
    DPLOG_L_ALERT = 5,
} dplog_level_e;

int64_t dplog_nowts(dplog_tsacc_e ta);
const struct tm* dplog_nowtm();

bool dplog_init(const char* file, dplog_level_e level, dplog_tsacc_e ta);
void dplog_print(const char* fmt, ...);
void dplog_write(dplog_level_e level, const char* domain, const char* fmt, ...);
void dpelog_write(dplog_level_e level, const char* fmt, ...);
dplog_level_e dplog_curlevel();
const char* dplog_curlname();
dplog_tsacc_e dplog_curtsacc();
void dplog_setlevel(dplog_level_e l);
void dplog_settsacc(dplog_tsacc_e ta);

void dplog_debug(const char* domain, const char* fmt, ...);
void dplog_info(const char* domain, const char* fmt, ...);
void dplog_notice(const char* domain, const char* fmt, ...);
void dplog_warn(const char* domain, const char* fmt, ...);
void dplog_error(const char* domain, const char* fmt, ...);
void dplog_alert(const char* domain, const char* fmt, ...);

void dpelog_debug(const char* fmt, ...);
void dpelog_info(const char* fmt, ...);
void dpelog_notice(const char* fmt, ...);
void dpelog_warn(const char* fmt, ...);
void dpelog_error(const char* fmt, ...);
void dpelog_alert(const char* fmt, ...);
]])

local C = ffi.C
local select = select
local string_format = string.format

local M = {}

local function _fmt(fmt, ...)
    if select("#", ...) == 0 then
        return fmt
    end
    return string_format(fmt, ...)
end

function M.debug(domain, fmt, ...)
    C.dplog_debug(domain, _fmt(fmt, ...))
end

function M.info(domain, fmt, ...)
    C.dplog_info(domain, _fmt(fmt, ...))
end

function M.notice(domain, fmt, ...)
    C.dplog_notice(domain, _fmt(fmt, ...))
end

function M.warn(domain, fmt, ...)
    C.dplog_warn(domain, _fmt(fmt, ...))
end

function M.error(domain, fmt, ...)
    C.dplog_error(domain, _fmt(fmt, ...))
end

function M.alert(domain, fmt, ...)
    C.dplog_alert(domain, _fmt(fmt, ...))
end

function M.e_debug(fmt, ...)
    C.dpelog_debug(_fmt(fmt, ...))
end

function M.e_info(fmt, ...)
    C.dpelog_info(_fmt(fmt, ...))
end

function M.e_notice(fmt, ...)
    C.dpelog_notice(_fmt(fmt, ...))
end

function M.e_warn(fmt, ...)
    C.dpelog_warn(_fmt(fmt, ...))
end

function M.e_error(fmt, ...)
    C.dpelog_error(_fmt(fmt, ...))
end

function M.e_alert(fmt, ...)
    C.dpelog_alert(_fmt(fmt, ...))
end

M.L_DEBUG = 0
M.L_INFO = 1
M.L_NOTICE = 2
M.L_WARN = 3
M.L_ERROR = 4
M.L_ALERT = 5

M.TA_SECOND = C.DPLOG_TA_SECOND
M.TA_MILLIS = C.DPLOG_TA_MILLIS
M.TA_MICROS = C.DPLOG_TA_MICROS

function M.init(file, level_, ta_)
    level_ = ffi.cast("dplog_level_e",
        tonumber(level_) or tonumber(C.dplog_curlevel()))
    ta_ = ffi.cast("dplog_tsacc_e",
        tonumber(ta_) or tonumber(C.dplog_curtsacc()))
    return C.dplog_init(tostring(file), level_, ta_)
end

function M.print(fmt, ...)
    C.dplog_print(_fmt(fmt, ...))
end

function M.level(level, domain, fmt, ...)
    C.dplog_write(ffi.cast("dplog_level_e", tonumber(level)), domain,
        _fmt(fmt, ...))
end

function M.e_level(level, fmt, ...)
    C.dpelog_write(ffi.cast("dplog_level_e", tonumber(level)), _fmt(fmt, ...))
end

function M.curlevel()
    return tonumber(C.dplog_curlevel())
end

function M.curlname()
    local s = C.dplog_curlname()
    return s == nil and nil or ffi.string(s)
end

function M.curtsacc()
    return tonumber(C.dplog_curtsacc())
end

function M.setlevel(level)
    C.dplog_setlevel(ffi.cast("dplog_level_e", tonumber(level)))
end

function M.settsacc(ta)
    C.dplog_settsacc(ffi.cast("dplog_tsacc_e", tonumber(ta)))
end

function M.timestamp(ta_)
    ta_ = ffi.cast("dplog_tsacc_e", tonumber(ta_) or M.TA_MILLIS)
    return tonumber(C.dplog_nowts(ta_))
end

function M.now(fmt_)
    local stm = C.dplog_nowtm()
    if stm == nil then
        return nil
    end
    if fmt_ == nil then
        return os.date("%c", os.time({
            year = stm.tm_year + 1900,
            month = stm.tm_mon + 1,
            day = stm.tm_mday,
            hour = stm.tm_hour,
            min = stm.tm_min,
            sec = stm.tm_sec,
        }))
    end
    if fmt_ == "*t" then
        return {
            sec = stm.tm_sec,
            min = stm.tm_min,
            hour = stm.tm_hour,
            day = stm.tm_mday,
            month = stm.tm_mon + 1,
            year = stm.tm_year + 1900,
            wday = stm.tm_wday + 1,
            yday = stm.tm_yday + 1,
            isdst = stm.tm_isdst == 1,
        }
    end
    return os.date(fmt_, os.time({
        year = stm.tm_year + 1900,
        month = stm.tm_mon + 1,
        day = stm.tm_mday,
        hour = stm.tm_hour,
        min = stm.tm_min,
        sec = stm.tm_sec,
    }))
end

return M

File: lib/dpapp/dplog.c
#include "dpapp/dplog.h"
#include <fcntl.h>
#include <stdarg.h>
#include <stdio.h>
#include <string.h>
#include <sys/time.h>
#include <unistd.h>

#define DPLOG_BUFSZ 2048

extern int dpevp_id();

static inline int _iclamp(int v, int l, int h)
{
    int t = v < l ? l : v;
    return t < h ? t : h;
}

typedef struct
{
    dplog_level_e _level;
    int _fd;
    dplog_tsacc_e _ta; // 时间精度：0=秒，1=毫秒，2=微秒
} dplog_t;

static dplog_t* _dplog_instance()
{
    static dplog_t logger = {._level = DPLOG_L_NOTICE, ._fd = -1, ._ta = 0};
    return &logger;
}

void dplog_print(const char* fmt, ...)
{
    dplog_t* logger = _dplog_instance();

    va_list args;
    va_start(args, fmt);
    char txt[DPLOG_BUFSZ] = "";

    int sz = vsnprintf(txt, DPLOG_BUFSZ, fmt, args);
    if (sz < DPLOG_BUFSZ) {
        sz++;
    } else {
        sz = DPLOG_BUFSZ;
    }
    txt[sz - 1] = '\n';

    ssize_t wsz = 0;
    if (logger->_fd >= 0) {
        wsz = write(logger->_fd, txt, sz);
    } else {
        wsz = write(fileno(stdout), txt, sz);
    }
    if (wsz < sz) {
        // 无操作
    }

    va_end(args);
}

void dplog_vprint(dplog_level_e l, const char* domain, const char* fmt, va_list arg)
{
    dplog_t* logger = _dplog_instance();
    if ((int)l < (int)logger->_level || fmt == NULL) {
        return;
    }

    const struct tm* ntm = dplog_nowtm();

    char txt[DPLOG_BUFSZ] = "";
    int tsz = 0;
    tsz = strftime(txt, DPLOG_BUFSZ, "%FT%T", ntm);

    if (logger->_ta != 0) {
        int64_t ts = dplog_nowts(logger->_ta);
        if (logger->_ta == 1) {
            tsz += snprintf(txt + tsz, DPLOG_BUFSZ - tsz, ".%03ld", ts % 1000);
        } else {
            tsz += snprintf(txt + tsz, DPLOG_BUFSZ - tsz, ".%06ld", ts % 1000000);
        }
    }

    int id = dpevp_id();
    if (id < 0) {
        id = 0;
    }

    if (domain) {
        tsz += snprintf(txt + tsz, DPLOG_BUFSZ - tsz, "[%03d-%s](%s) ", id,
            dplog_sname(l), domain);
    } else {
        tsz += snprintf(txt + tsz, DPLOG_BUFSZ - tsz, "[%02d-%s] ", id,
            dplog_sname(l));
    }
    int sz = vsnprintf(txt + tsz, DPLOG_BUFSZ - tsz, fmt, arg);
    if (sz >= DPLOG_BUFSZ - tsz) {
        tsz = DPLOG_BUFSZ - 1;
    } else {
        tsz += sz;
    }
    txt[tsz] = '\n';
    tsz++;

    ssize_t wsz = 0;
    if (logger->_fd >= 0) {
        wsz = write(logger->_fd, txt, tsz);
    } else {
        wsz = write(fileno(stdout), txt, tsz);
    }
    if (wsz < tsz) {
        // 无操作
    }
}

dplog_level_e dplog_curlevel()
{
    return _dplog_instance()->_level;
}

dplog_tsacc_e dplog_curtsacc()
{
    return _dplog_instance()->_ta;
}

const char* dplog_curlname()
{
    return dplog_lname(dplog_curlevel());
}

void dplog_setlevel(dplog_level_e l)
{
    _dplog_instance()->_level = _iclamp(l, 0, 5);
}

void dplog_settsacc(dplog_tsacc_e ta)
{
    _dplog_instance()->_ta = (dplog_tsacc_e)_iclamp(ta, 0, 2);
}

const char* dplog_lname(dplog_level_e l)
{
    static const char* names[6] = {"debug", "info", "notice", "warning", "error",
        "alert"};
    return names[_iclamp(l, 0, 5)];
}

dplog_level_e dplog_namel(const char* n)
{
    if (strcasecmp(n, "debug") == 0) {
        return DPLOG_L_DEBUG;
    } else if (strcasecmp(n, "info") == 0) {
        return DPLOG_L_INFO;
    } else if (strcasecmp(n, "notice") == 0) {
        return DPLOG_L_NOTICE;
    } else if (strcasecmp(n, "warning") == 0) {
        return DPLOG_L_WARN;
    } else if (strcasecmp(n, "error") == 0) {
        return DPLOG_L_ERROR;
    } else if (strcasecmp(n, "alert") == 0) {
        return DPLOG_L_ALERT;
    } else {
        return DPLOG_L_ALERT;
    }
}

const char* dplog_sname(dplog_level_e l)
{
    static const char* names[6] = {"D", "I", "N", "W", "E", "A"};
    return names[_iclamp(l, 0, 5)];
}

bool dplog_init(const char* file, dplog_level_e level, dplog_tsacc_e ta)
{
    if (file == NULL) {
        file = "/dev/stdout";
    }
    dplog_t* logger = _dplog_instance();
    logger->_level = _iclamp((int)level, 0, 5);
    logger->_ta = (dplog_tsacc_e)_iclamp(ta, 0, 2);

    int fd = open(file, O_WRONLY | O_APPEND | O_CREAT, 644);
    if (fd >= 0) {
        if (logger->_fd >= 0) {
            close(logger->_fd);
        }
        logger->_fd = fd;
        return true;
    } else {
        return false;
    }
}

#define LOG(l)                                                                      \
    if ((int)l < (int)dplog_curlevel()) {                                           \
        return;                                                                     \
    }                                                                               \
    va_list args;                                                                   \
    va_start(args, fmt);                                                            \
    dplog_vprint(l, domain, fmt, args);                                             \
    va_end(args);

void dplog_write(dplog_level_e level, const char* domain, const char* fmt, ...)
{
    LOG(level);
}

void dplog_debug(const char* domain, const char* fmt, ...)
{
    LOG(DPLOG_L_DEBUG);
}
void dplog_info(const char* domain, const char* fmt, ...)
{
    LOG(DPLOG_L_INFO);
}
void dplog_notice(const char* domain, const char* fmt, ...)
{
    LOG(DPLOG_L_NOTICE);
}
void dplog_warn(const char* domain, const char* fmt, ...)
{
    LOG(DPLOG_L_WARN);
}
void dplog_error(const char* domain, const char* fmt, ...)
{
    LOG(DPLOG_L_ERROR);
}
void dplog_alert(const char* domain, const char* fmt, ...)
{
    LOG(DPLOG_L_ALERT);
}

#define ELOG(l)                                                                     \
    if (l < dplog_curlevel()) {                                                     \
        return;                                                                     \
    }                                                                               \
    va_list args;                                                                   \
    va_start(args, fmt);                                                            \
    dplog_vprint(l, NULL, fmt, args);                                               \
    va_end(args);

void dpelog_write(dplog_level_e level, const char* fmt, ...)
{
    ELOG(level);
}

void dpelog_debug(const char* fmt, ...)
{
    ELOG(DPLOG_L_DEBUG);
}
void dpelog_info(const char* fmt, ...)
{
    ELOG(DPLOG_L_INFO);
}
void dpelog_notice(const char* fmt, ...)
{
    ELOG(DPLOG_L_NOTICE);
}
void dpelog_warn(const char* fmt, ...)
{
    ELOG(DPLOG_L_WARN);
}
void dpelog_error(const char* fmt, ...)
{
    ELOG(DPLOG_L_ERROR);
}
void dpelog_alert(const char* fmt, ...)
{
    ELOG(DPLOG_L_ALERT);
}

struct dplog_time
{
    long nowSec;
    struct tm nowTm;
};

static __thread struct dplog_time _gTime = {.nowSec = 0};

const struct tm* dplog_nowtm()
{
    time_t t = time(NULL);
    if (t > _gTime.nowSec) {
        localtime_r(&t, &_gTime.nowTm);
        _gTime.nowSec = t;
    }
    return &_gTime.nowTm;
}

int64_t dplog_nowts(dplog_tsacc_e ta)
{
    struct timeval now = {0};
    gettimeofday(&now, NULL);
    unsigned long long u = now.tv_sec;

    switch (ta) {
    case DPLOG_TA_SECOND:
        return u;
    case DPLOG_TA_MILLIS:
        u *= 1000;
        u += now.tv_usec / 1000;
        return u;
    case DPLOG_TA_MICROS:
        u *= 1000000;
        u += now.tv_usec;
        return u;
    default:
        u *= 1000;
        u += now.tv_usec / 1000;
        return u;
    }
}
